### ISSUE TITLE
HiDDeN decoder train

### ISSUE THREAD

**Issue by jang-1996** [facebookresearch/stable_signature#41]
Hi!
Thank you for the awesome work.

I’m training the decoder in the 32-bit setting with rotation, crop, and JPEG distortions on a single GPU, but my bit accuracy plateaus around ~80%. 

I’d really appreciate any guidance.

Augmentation (rotation snippet)

```
aff1 = K.RandomAffine(degrees=(-45, 45), p=1)
aff2 = K.RandomAffine(degrees=(30, 30), p=1)
aff3 = K.RandomAffine(degrees=(-30, -30), p=1)
augmentations += [aff1]
augmentations += [aff2, aff3]

```

Run command

```
torchrun --nproc_per_node=1 main.py --dist False \
  --val_dir /home/jang/stable_signature/hidden/dataset/val5000 \
  --train_dir /home/jang/stable_signature/hidden/dataset/train10000 \
  --output_dir output_32 --eval_freq 50 --img_size 256 --num_bits 32 \
  --batch_size 32 --epochs 500 \
  --optimizer Lamb,lr=5e-3 \
  --scheduler CosineLRScheduler,lr_min=1e-6,t_initial=500,warmup_lr_init=1e-6,warmup_t=5 \
  --p_color_jitter 0.0 --p_blur 0.0 --p_rot 1.0 --p_crop 1.0 --p_res 0.0 --p_jpeg 1.0 \
  --scaling_w 0.3 --scale_channels False --attenuation none \
  --loss_w_type bce --loss_margin 1

```

What I tried these following [the solved issue](https://github.com/facebookresearch/stable_signature/issues/5#issuecomment-1790640271) 

- Tuning the learning rate

- Removing lr_mult = params.batch_size * utils.get_world_size() / 512.0

Questions

- For 32-bit decoding under these distortions, are there recommended LR / scheduler settings to push accuracy over 99%?

Thanks again!

**Comment by pierrefdz** [comment]
Hi, I think the issue comes from the lamb optimizer. I remember that in my case the lamb optimizer was surprisingly good at the very end of the training (you can plot the training logs in the ones I think I gave). Maybe you can change to Adam if you've already tried changing the LR.

Besides, another difference might be that in my case, since I have several gpus and that each rank gets a different seed, the backprop is done on several augs every time. In your case, with 1 gpu, all images in the batch have same augmentation and the backprop is done only with this one. What you can try: (i) apply the augmentation differently depending on the image of the batch (but be careful of crops since it means that tensors cannot be batched after augmentation; (ii) do gradient accumulation and simulate more GPUs.

If you feel like it's taking too long, you can also remove some augmentations
